### PR TITLE
[AIRFLOW-3602] Improve ImapHook handling of retrieving no attachments

### PR DESCRIPTION
**Pull request by feluelle**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3602
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR changes the ImapHook so it will fail when no attachments have been found.
It also:
- fixes escaping chars and symlink tests not testing correctly
- adds note for handling symlinks

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by feluelle** [comment]
PTAL @kaxil I would really like to have this in 1.10.2 if possible. Or what do you actually think of this PR?

**Comment by kaxil** [comment]
I think we can raise a warning but we shouldn't raise an exception.

**Comment by feluelle** [comment]
But then the task won't fail if there is no attachment.
And how do you know then if, for example, the transfer for this attachment works. I don't think it belongs to the transfer operator to check whether there is an attachment or not.

**Comment by feluelle** [comment]
It is similar to the S3Hook if you want to download a file from s3 you get an exception when there isn't a file with the given name. I think the same goes for a mail server and downloading its mail attachments. If you want to be sure that there is an attachment with the given name you can run a sensor task (`ImapAttachmentSensor`) before the actual download.

**Comment by feluelle** [comment]
@Fokko @feng-tao @ashb what do you guys think of this?

**Comment by ashb** [comment]
My first though is that the Hook should probably just return `[]` (no attacments) and this sort of error handling/exception should be thrown from the Operator.

The S3 case is different - getting the attachments of a message is closer to listing keys (which doesn't throw any exceptions, it returns an empty list) than getting a specific file.

**Comment by feluelle** [comment]
> The S3 case is different - getting the attachments of a message is closer to listing keys (which doesn't throw any exceptions, it returns an empty list) than getting a specific file.

Is it really? But the `download_mail_attachments` actually also raises no exception atm because to create files from the attachments it loops through the attachments that have been found. So also this task will succeed even it hasn't downloaded anything.

**Comment by ashb** [comment]
```bash
$ mkdir tmp
$ ls tmp/
$ echo $?
0
```

**Comment by feluelle** [comment]
Hey @ashb & @kaxil,

please take a look once more. I changed it so you are able to specify what should happen when there is no attachment with the given name.

File: airflow/contrib/hooks/imap_hook.py [hunk 1]
```diff
@@ -55,72 +53,98 @@ def has_mail_attachment(self, name, mail_folder='INBOX', check_regex=False):
...
+
         self._create_files(mail_attachments, local_output_directory)
 
-    def _retrieve_mails_attachments_by_name(self, name, mail_folder, check_regex,
-                                            latest_only):
+    def _handle_not_found_mode(self, not_found_mode):
+        if not_found_mode is 'raise':
+            raise AirflowException('No mail attachments found!')
+        if not_found_mode is 'warn':
+            self.log.warning('No mail attachments found!')
```
Comment (ashb) [inline airflow/contrib/hooks/imap_hook.py:147]: I'd like a 'silent' option - do nothing at all. If I'm conditionally handling this in my custom operator I don't want a log message.

Comment (feluelle) [reply airflow/contrib/hooks/imap_hook.py:147]: Good Point! Will add this. :)

Comment (feluelle) [reply airflow/contrib/hooks/imap_hook.py:147]: Added silent mode (`not_found_mode='ignore'`) and rebased onto current master

**Review by ashb — APPROVED** [review]

**Comment by ashb** [comment]
Waiting for green tests then will merge

**Comment by kaxil** [comment]
Thanks @feluelle

**Pull request merged**